Fix TensorExpressions::evaluate to use LHS index order for LHS Tensor

macedo22 opened the pull request:
## Proposed changes

Given a right hand side (RHS) tensor, its generic index order, and a left hand side (LHS) generic index order, `TensorExpressions::evaluate` constructs a left hand side (LHS) tensor with the desired LHS generic index order. Currently, `TensorExpressions::evaluate` constructs this LHS tensor using the RHS symmetry and list of indices, but it should be constructed using the (potentially reordered) LHS symmetry and list of indices, based on the desired LHS generic index order. This PR fixes this by determining the (potentially reordered) LHS symmetry and list of indices and constructing the LHS tensor using them. In addition, a unit test for `TensorExpressions::evaluate` was added.

### Types of changes:

- [x] Bugfix
- [ ] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

If we have a `Tensor` named `R` and a `TensorExpression` named `R_abc`:
```
  Tensor<double, Symmetry<2, 1, 2>,
         index_list<SpacetimeIndex<3, UpLo::Lo, Frame::Grid>,
                    SpacetimeIndex<4, UpLo::Lo, Frame::Grid>,
                    SpacetimeIndex<3, UpLo::Lo, Frame::Grid>>>
      R{};
  auto R_abc = R(ti_a, ti_b, ti_c);
```
and then evaluate the RHS `TensorExpression` to a LHS `Tensor` named `L_bac`, where the first two indices switched:
```
  auto L_bac = TensorExpressions::evaluate<ti_b_t, ti_a_t, ti_c_t>(R_abc);
```
`L_bac`'s symmetry should be reordered according to swapping the first two indices. The RHS's symmetry is `Symmetry<2, 1, 2>`. If we reordered `<2, 1, 2>` to match this swap, then the reordered LHS symmetry should be `Symmetry<1, 2, 2>`. Likewise, the indices in the `index_list` of the LHS `Tensor` should also be reordered according to swapping the first two indices. In order words, the LHS `index_list` should be:
```
index_list<SpacetimeIndex<4, UpLo::Lo, Frame::Grid>,
           SpacetimeIndex<3, UpLo::Lo, Frame::Grid>,
           SpacetimeIndex<3, UpLo::Lo, Frame::Grid>>
```
Currently, `evaluate` uses the RHS symmetry's ordering and index list ordering for constructing the LHS `Tensor`. This PR fixes is it to use the LHS symmetry and index list ordering, as described above.

In the new unit test, `TensorExpressions::evaluate` is tested using rank 0 through rank 4 `TensorExpression`s as the argument. The test checks that evaluating a RHS `TensorExpression` to a LHS `Tensor` with some RHS generic index order and desired LHS generic index order yields a LHS `Tensor` whose indices and components have been properly ordered (potentially reordered). This test iterates over RHS `Tensor`s with different index dimensions, index types (`SpatialIndex`, `SpacetimeIndex`), frame types (`Grid`, `Inertial`), and symmetries using different RHS and LHS generic index orders.

macedo22 committed:
Fix evaluate by constructing LHS tensor with LHS index order

Add single tensor tests for evaluate

Change Datatype to DataType

Cosmetic updates

Use explicit types for tested evaluated LHS tensors

Add missing two rank 2 nonsymmetric test cases

Rename LhsTensor -> LhsTensorSymmAndIndices

Remove redundant template params for symmetric test functions

nilsdeppe commented on tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank0TestHelpers.hpp:
```diff
@@ -0,0 +1,24 @@
+#include "DataStructures/Tensor/Expressions/TensorExpression.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+
+/// \ingroup TestingFrameworkGroup
```
Can you put this inside the namespace `TestHelpers::TensorExpressions` please?

macedo22 replied:
Done

nilsdeppe commented on tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank1TestHelpers.hpp:
```diff
@@ -0,0 +1,74 @@
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \ingroup TestingFrameworkGroup
```
Can you put these (all the files) into `TestHelpers::TensorExpressions` please?

nilsdeppe commented on tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank0TestHelpers.hpp:
```diff
@@ -0,0 +1,24 @@
+/// \tparam Datatype the type of data being stored in the Tensors
+/// \param data the data being stored in the Tensors
+template <typename Datatype>
+void test_evaluate_rank_0(const Datatype& data) {
```
can you mark these all `noexcept`?

nilsdeppe commented on tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank2TestHelpers.hpp:
```diff
@@ -0,0 +1,177 @@
+void test_evaluate_rank_2_core(
+    const TensorIndexA& tensorindex_a, const TensorIndexB& tensorindex_b) {
+  Tensor<Datatype, RhsSymmetry, RhsTensorIndexTypeList> R_ab{};
+  std::iota(R_ab.begin(), R_ab.end(), 0.0);
```
I'm not sure this initialization is correct for `DataType == DataVector`. You need to construct the `DataVector`s with a certain size. We usually do this using
```cpp
auto R_ab = make_with_value<Tensor<Datatype, RhsSymmetry, RhsTensorIndexTypeList>>(0.0, used_for_size);
```
where `used_for_size` is a `DataType` object. Eg a `double` or `DataVector` of the size you want. Nowadays you might be able to just do:
```cpp
auto R_ab = make_with_value<Tensor<Datatype, RhsSymmetry, RhsTensorIndexTypeList>>(0.0, 5_st);
```
and include `Utilities/Literals.hpp`

This applies everywhere you construct tensors by default

macedo22 replied:
@nilsdeppe Thanks! I saw some changes with `make_with_value` and `Tensor.hpp` in the last couple weeks, so I took care to rebase before I tried this. I tried rebasing and then replacing this line:

```
Tensor<Datatype, RhsSymmetry, RhsTensorIndexTypeList> R_ab{};
```

with

```
auto R_ab = make_with_value<Tensor<Datatype, RhsSymmetry, RhsTensorIndexTypeList>>(0.0, 5_st);
```

When I do this, [this assert](https://github.com/sxs-collaboration/spectre/blob/49b82942457aba8deb01f65edc2e3667841c4fed/src/Utilities/MakeWithValue.hpp#L31) fails. I then dug around in `Test_Tensor.cpp` for non-default `Tensor` instantiation with `DataVector`s and saw [this](https://github.com/sxs-collaboration/spectre/blob/49b82942457aba8deb01f65edc2e3667841c4fed/tests/Unit/DataStructures/Tensor/Test_Tensor.cpp#L1056) (for example). So, when I instead swapped out the original line for this:

```
Tensor<Datatype, RhsSymmetry, RhsTensorIndexTypeList> R_ab(5_st);
```

it compiled fine and passed the tests. I added a couple new tests (locally) with `DataVector`s and those pass, as well, such as:

```
TestHelpers::TensorExpressions::test_evaluate_rank_2_no_symmetry<
      DataVector, SpacetimeIndex, SpacetimeIndex, UpLo::Lo, UpLo::Up>(
          ti_f, ti_G);
```

Would this change be alright, instead? From my understanding, this would mean that when `Datatype == DataVector` these two lines (the first one changed from the original):

```
Tensor<Datatype, RhsSymmetry, RhsTensorIndexTypeList> R_ab(5_st);
std::iota(R_ab.begin(), R_ab.end(), 0.0);
```

would make the elements of `R_ab` be size 5 `DataVector`s, where the first `DataVector` is `(0, 0, 0, 0, 0)`, the second is `(1, 1, 1, 1, 1)`, the third `(2, 2, 2, 2, 2)`, and so forth. The only comment I have on this is that I don't know how much it matters to have a test where the elements of an individual `DataVector` are not all the same value, as they would be with this suggested change.

nilsdeppe replied:
Yep, sounds good to me! :)

nilsdeppe commented on tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank3TestHelpers.hpp:
```diff
@@ -0,0 +1,444 @@
+#undef DIM_AC
+}
+
+/// \ingroup TestingFrameworkGroup
```
I would say you don't need to document all these different slightly different variations of this function in such detail. There are a couple options
- Only use `\brief`
- Use Doxygens `// @{` and `// @}` to apply the same comment to all the functions without having to copy-paste or write it out again
- Use Doxygen's `copydoc` (I think that's the right now) function to just copy the doc automatically between the different functions.

macedo22 replied:
@nilsdeppe If I remove the `\details` from `test_evaluate_rank_3_no_symmetry`, `test_evaluate_rank_3_ab_symmetry`, `test_evaluate_rank_3_ac_symmetry`, `test_evaluate_rank_3_bc_symmetry`, and `test_evaluate_rank_3_abc_symmetry` I think the the issue still remains that I couldn't use `copydoc` to wholesale copy the documentation around because the `\brief` between them differs. The `\brief` descriptions differ only slightly, but I think the way they differ is what's important to point out - that they are testing different symmetries - because otherwise it might not be clear why there are multiple functions defined that look almost the same. That's just my take though! Honestly, seeing the difference in the names of the functions, alone, might be enough to extract that information. Meaning, maybe I could just have something like this be the shared `\brief` between all and not mention anything about symmetry at all:
```
/// \brief Iterate testing of evaluating single rank 3 Tensors on multiple Frame
/// types and dimension combinations
```
 If you think they can all have the same `\brief` and call it a day, then I think I can simply make this one shared `\brief` and then use `copydoc` and call it a day, because the `\details`, `\tparam`s, and `param`s are what are the same between all of these - only their `\briefs` differ. I saw that there is a way to copy over just the `\details` from another function, but I don't see a way to copy over just the `\tparam`s and `\param`s. Ideally, it would be nice if I could leave each of the `\brief`s and then copy over the `\details` with `\copydetails` and the `\tparam`s and `\param`s with a similar function. Do you know if it's possible to do something like this? The only possible thing I have found online would be to define a separate file with the text to copy over.

So I guess the tldr is: which do you think is better and feasible?
1) changing the `\brief`s of those 5 to be exactly the same (no mention of symmetry, like in the example `\brief` I gave above), and then use `copydoc` to copy the entire documentation (`\brief`, `\details`, `param`s, and `\tparam`s) to be exactly the same between the 5
2) keeping the `\brief`s different, but copying the `\details`, `tparam`s, and `param`s (if that's even possible) between the 5
3) removing `\details` altogether, which still leaves the `\tparam`s and `\param`s as something chunky that is manually copied around between the 5 functions

nilsdeppe replied:
So what we've done in other situations like this where basically several functions are nearly identical with possibly long comments we have a single comment along the lines of:
```cpp
/// \brief Iterate testing of evaluating single rank 3 Tensors on multiple Frame
/// types and dimension combinations
///
/// We test various different symmetries to ensure that the code works correctly in with
/// symmetries. The symmetries we test are:
/// - A
/// - B
/// - C
```
You can also list other things that are checked, such as different dimensions, and/or valence, and/or spacetime vs spatial index.

Regarding the `\tparam` and `\param`. What we've done here is basically give a list of the parameters without using these keywords. Since Doxygen checks the keywords it becomes a bit tedious to handle various permutations. I would also say you do something like

```
Check with indices A and B that have `ValenceA` and `ValanceB`, `TensorIndexA` and `TensorIndexB` (either spatial or spacetime)
```
Having said that, my view in general is that deduced template parameters don't really need documenting (they are deduced based off the value passed in), while explicit template parameters should be documented (eg DataType).

Having said all this: what if you eliminate the documentation for the `core` functions (they are really the interface and could/should go into a detail namespace), and then trim the documentation for eg `test_evaluate_rank_2_symmetric` to:
```cpp
/// \ingroup TestingFrameworkGroup
/// \brief Iterate testing of evaluating a single rank 2 Tensors on multiple
/// Frame types and dimension combinations for symmetric indices
///
/// \details `TensorIndexA` and `TensorIndexB` can be any type of TensorIndex
/// and are not necessarily `ti_a_t` and `ti_b_t`. The "A" and "B" suffixes just
/// denote the ordering of the generic indices of the RHS tensor expression. In
/// the RHS tensor expression, it means `TensorIndexA` is the first index used
/// and `TensorIndexB` is the second index used.
```
I think that gives all the details you add with the `\tparam` and `\param` specifications. What do you think?

macedo22 replied:
@nilsdeppe Thanks so much for your input! I have some comments and questions that became a bit lengthy, so please ping me on slack if it's easier to discuss the things below over slack. Here are my comments:

1) Regarding this suggestion you made:

```
/// \brief Iterate testing of evaluating single rank 3 Tensors on multiple Frame
/// types and dimension combinations
///
/// We test various different symmetries to ensure that the code works correctly in with
/// symmetries. The symmetries we test are:
/// - A
/// - B
/// - C
```

I want to make sure I understand correctly that you mean for this to be something written once and copied to the other functions for different symmetries (e.g. write it once as the documentation for `test_evaluate_rank_3_no_symmetry`, then use `\copydoc` to copy it to the documentation for `test_evaluate_rank_3_ab_symmetry`, `test_evaluate_rank_3_ac_symmetry`, `test_evaluate_rank_3_bc_symmetry`, and `test_evaluate_rank_3_abc_symmetry`). If this is the case:
- I think it reads a little strangely to have "we test different symmetries, here is the list" as part of the documentation within each of these instead of that being stated at a more outer / upper level of documentation for something that contains these functions, because when it's read, it almost sounds like it means "this function tests the following symmetries," but it only tests one of them. But I'm also not strongly opposed to this suggestion if this is the best way to keep things clean and concise. Plus, I can try to think of a way to reword it so it's clear that although we're giving a list of symmetries, that the current function only tests one of them.
- If we were to do this, this solves the issue of repeated documentation for the functions that do the same thing for different symmetries. That group of functions currently has the exact same documentation for the `\details`s, `\param`s, and `\tparam`s, but their `\brief`s differ only slightly in wording due to them dealing with different symmetries. Therefore, it's my opinion that if we do this for something like these 5 functions, I can just write something similar to what this snippet above is and copy it over to the other four functions, but still have that documentation have the `\tparam`s and `\param`s (plus deduced `\tparam`s removed) because those are the same across all 5 functions. This is essentially the option (1) that I mentioned in my previous comment, but with your comments above on testing different symmetries.
- If I'm misunderstanding this suggestion, let me know. Or if you can point me to an example of where we do this elsewhere in the code, let me know

2) Regarding this suggestion you made:

```
/// \ingroup TestingFrameworkGroup
/// \brief Iterate testing of evaluating a single rank 2 Tensors on multiple
/// Frame types and dimension combinations for symmetric indices
///
/// \details `TensorIndexA` and `TensorIndexB` can be any type of TensorIndex
/// and are not necessarily `ti_a_t` and `ti_b_t`. The "A" and "B" suffixes just
/// denote the ordering of the generic indices of the RHS tensor expression. In
/// the RHS tensor expression, it means `TensorIndexA` is the first index used
/// and `TensorIndexB` is the second index used.
```

- This suggestion makes me think I'm misunderstanding what is being called redundant. I had thought that you meant the documentation between something like the following was redundant: `test_evaluate_rank_3_no_symmetry`, `test_evaluate_rank_3_ab_symmetry`, `test_evaluate_rank_3_ac_symmetry`, `test_evaluate_rank_3_bc_symmetry`, and `test_evaluate_rank_3_abc_symmetry`. If this is the case, then I think the problem I mentioned in my previous comment remains - that I can't use something like `\copydoc` to copy around documentation like this between these 5 functions because the `\brief` above is unique to the function due to it specifying which symmetry it deals with (and thus can't be copied around). If I have a misunderstanding of what is being called redundant, let me know!
- I of course will go with whatever is requested of me by you and the other developers, but I personally don't think we should eliminate the documentation for the `core` functions altogether because although they contain the innermost details of the testing, I think it's worth pointing out what these are doing because callers like `test_evaluate_rank_3_ac_symmetry` or its documentation don't capture what `test_evaluate_rank_3_core` is doing. The callers of these `core` functions merely generate a multitude of tensors as test "input," but I think it's worth explaining in the documentation for the `core` functions that these `core` functions are checking that the data is correct when these "input" tensors are evaluated with different LHS tensor index orders, which (in my opinion) is probably the most important detail to mention about the unit test. If you think there is a place to mention this detail that is better or perhaps more consistent with what is done elsewhere in SpECTRE, let me know! I just think it's important to mention that the data is being checked when evaluating with all the LHS index combinations. To me, that place makes sense in the documentation for the `core` functions, but maybe there is a better spot for it!

Thanks again for the suggestions and input!

nilsdeppe replied:
1. I personally like the style in your example. I agree that the wording feels a bit awkward since the function does only one of the symmetries. What about something like "This function tests one of the following symmetries:.... Other functions test the additional symmetries"?
2. That's fair. So something for you to decide then is: how much should be described is doxygen documentation vs. should just be a C++ comment at the start of the or interspersed throughout the function definition. I think this is a matter of taste mostly, so decide something and let's go with that :)

macedo22 replied:
@nilsdeppe 

1. Done!
2. I went ahead and left a decent amount of documentation. If you think the inline C++ comments in the `core` functions are cluttering or too redundant with their `\details`, let me know and I can remove one of them. Also, if you think this part shouldn't be in both the `core` functions and their callers, let me know which one you think I should remove:
 
```
/// \details `TensorIndexA` and `TensorIndexB` can be any type of TensorIndex
/// and are not necessarily `ti_a_t` and `ti_b_t`. The "A" and "B" suffixes just
/// denote the ordering of the generic indices of the RHS tensor expression. In
/// the RHS tensor expression, it means `TensorIndexA` is the first index used
/// and `TensorIndexB` is the second index used.
```

I left the `\details`, `\tparam`s, and `\param`s for the `core` functions as-is

macedo22 commented on tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank2TestHelpers.hpp:
```diff
@@ -0,0 +1,177 @@
+          UpLo ValenceA, UpLo ValenceB, typename TensorIndexA,
+          typename TensorIndexB>
+void test_evaluate_rank_2_no_symmetry(const TensorIndexA& tensorindex_a,
+                                      const TensorIndexB& tensorindex_b) {
```
@nilsdeppe I interpreted your other review comment to mean to add `noexcept` to all of the `test_evaluate_rank_#_core` and `test_evaluate_rank_0` functions (i.e. all the innermost functions that use `CHECK`), but do you also want `noexcept` used with *all* of the functions defined in these `EvaluateRank#TestHelpers.hpp` files, like these ones?

nilsdeppe replied:
Yep! Our general policy is "noexcept every function". The only exception (no pun intended) is the option parsing code

macedo22 commented:
@nilsdeppe The first commit I added (f94c9f3) just renames `Datatype` -> `DataType` to be consistent with other code in SpECTRE. The second commit I added (b15bed5) addresses all of the following in one commit: putting the test functions into the `TestHelpers::TensorExpressions` namespace, making the `evaluate` test functions `noexcept`, fixing the `Tensor` instantiation for handling `DataVector`s properly, adding test cases for `DataVector`s, fixing up the documentation according to our above conversation, and removing deduced template parameters from the test functions.

When we spoke earlier, you mentioned that one commit for addressing the change requests was okay, but if going through b15bed5 proves to be difficult due to the number of changes, let me know and I can resubmit these changes as multiple fixup commits.

nilsdeppe commented on tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank4TestHelpers.hpp:
```diff
@@ -12,6 +12,9 @@
 #include "DataStructures/Tensor/Tensor.hpp"
 #include "Utilities/TMPL.hpp"
 
+namespace TestHelpers {
```
clang-tidy will want you to do `TestHelpers::TensorExpressions`. This type of namespace scoping is a C++17 feature :)

macedo22 replied:
Done!

nilsdeppe commented on tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank3TestHelpers.hpp:
```diff
@@ -11,8 +11,12 @@
+#include "Utilities/Literals.hpp"
 #include "Utilities/TMPL.hpp"
 
+namespace TestHelpers {
```
clang-tidy will want you to do `TestHelpers::TensorExpressions`. This type of namespace scoping is a C++17 feature :)

nilsdeppe commented on tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank2TestHelpers.hpp:
```diff
@@ -13,6 +13,9 @@
 #include "Utilities/GenerateInstantiations.hpp"
 #include "Utilities/TMPL.hpp"
 
+namespace TestHelpers {
```
clang-tidy will want you to do `TestHelpers::TensorExpressions`. This type of namespace scoping is a C++17 feature :)

nilsdeppe commented on tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank1TestHelpers.hpp:
```diff
@@ -13,6 +13,9 @@
 #include "Utilities/GenerateInstantiations.hpp"
 #include "Utilities/TMPL.hpp"
 
+namespace TestHelpers {
```
clang-tidy will want you to do `TestHelpers::TensorExpressions`. This type of namespace scoping is a C++17 feature :)

nilsdeppe commented on tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank0TestHelpers.hpp:
```diff
@@ -7,18 +7,23 @@
 #include "DataStructures/Tensor/Expressions/TensorExpression.hpp"
 #include "DataStructures/Tensor/Tensor.hpp"
 
+namespace TestHelpers {
```
clang-tidy will want you to do `TestHelpers::TensorExpressions`. This type of namespace scoping is a C++17 feature :)

nilsdeppe commented on tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank1TestHelpers.hpp:
```diff
@@ -21,20 +24,22 @@
+  Tensor<DataType, Symmetry<1>, TensorIndexTypeList> R_a(5_st);
   std::iota(R_a.begin(), R_a.end(), 0.0);
 
+  /*std::cout << "Rank 1 elements\n\n";
```
I assume this is left over from debugging and can be remove :)

macedo22 replied:
Thanks for catching this!

nilsdeppe reviewed (commented):
This looks great! Please go ahead and squash the commits appropriately :)

macedo22 commented:
@nilsdeppe Updated and squashed!

nilsdeppe reviewed (dismissed):
LGTM! @fmahebert would you be willing to be the second reviewer on this?

fmahebert commented on tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank1TestHelpers.hpp:
```diff
@@ -0,0 +1,73 @@
+/// \param tensorindex the TensorIndex used in the the TensorExpression,
+/// e.g. `ti_a`
+template <typename DataType, typename TensorIndexTypeList, typename TensorIndex>
+void test_evaluate_rank_1_core(const TensorIndex& tensorindex) noexcept {
```
[consistency nitpick, feel free to ignore] I suggest renaming "_core" -> "_impl" as this is something we do often through the codebase. If you do make this change, make it also in the macros and in the higher-rank files

macedo22 replied:
Done!

fmahebert commented on tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank1TestHelpers.hpp:
```diff
@@ -0,0 +1,73 @@
+  Tensor<DataType, Symmetry<1>, TensorIndexTypeList> R_a(5_st);
+  std::iota(R_a.begin(), R_a.end(), 0.0);
+
+  const size_t dim = tmpl::at_c<TensorIndexTypeList, 0>::dim;
```
Please move this line down, to just above the for-loop where it is used. Also in the higher-rank cases.

macedo22 replied:
Done!

fmahebert commented on tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank2TestHelpers.hpp:
```diff
@@ -0,0 +1,153 @@
+  const size_t dim_b = tmpl::at_c<RhsTensorIndexTypeList, 1>::dim;
+
+  // L_{ab} = R_{ab}
+  const auto L_ab = ::TensorExpressions::evaluate<TensorIndexA, TensorIndexB>(
```
Would it make the test more rigorous if instead of `const auto L_ab = ...` you wrote `const Tensor<DataType, RhsSymmetry, tmpl::list<TensorIndexA, TensorIndexB>> L_ab = ...` ?

I'm thinking of a hypothetical bug where `TensorExpressions::evaluate` might return an incorrect tensor type... here you are checking values of components, but not the type of the returned object. I think in most cases an incorrect tensor type would fail somewhere in the component-value checks below... but why not make it fail sooner, here? Perhaps this is already tested elsewhere?

If you do make a change for this, make it in the other test cases too. But please just inform me of anything I'm missing that makes this redundant or too difficult (handling the symmetry in a general way?) :)

macedo22 replied:
I think this is a great suggestion. Last night I was working on the contraction code and I ran into a bug like the one you're mentioning, and this would've been caught earlier by my contraction tests if I hadn't been using `auto`, so today I was going to add asserts or use `CHECK` to check the indices on the LHS - but I like this much better. Thanks!

macedo22 replied:
Something I didn't notice at first though is that if I'm to do this for all of the cases in the tests, then instead of
```
const Tensor<DataType, RhsSymmetry, tmpl::list<TensorIndexA, TensorIndexB>> LHS = ...
```
it will have to be
```
const Tensor<DataType, <the LHS reordered symmetry>, <the LHS reordered tensor index type list>> LHS = ...
```
so that the LHS `Tensor` has its symmetry and tensor index type list reordered the same way as the `TensorIndex`s are being reordered (essentially, what the struct in `Evaluate.hpp` currently named `LhsTensor` does)

fmahebert replied:
Oh good! It might be helpful to add a short comment (just once per file the first time you replace `auto` with the expected type) saying something like `// Use explicit type (vs auto) so the compiler checks the return type of the TensorExpression` ? That might help avoid someone thinking they can simplify the code in the future by putting `auto` back in 😛

macedo22 replied:
Done! The code got decently more verbose for the rank 3 and rank 4 `impl` (previously, `core`) functions, so let me know if you would prefer me to handle this in a different way than I did in 317e87d

fmahebert replied:
I agree this did get a bit verbose, but it does the checks we wanted. I'm happy with it!

I also can't think of an alternative way of doing these checks that isn't equally verbose, because regardless of how you structure things, you're going to need to write out the expected symmetry and index list – and that's the most cumbersome part.

fmahebert commented on tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank2TestHelpers.hpp:
```diff
@@ -0,0 +1,153 @@
+    UpLo ValenceB, typename TensorIndexA, typename TensorIndexB>
+void test_evaluate_rank_2_symmetric(
+    const TensorIndexA& tensorindex_a,
+    const TensorIndexB& tensorindex_b) noexcept {
```
Might it be useful to add some ASSERTS that `tensorindex_a`, `tensorindex_b` can indeed be symmetric? That is to say, assert they've got the same dim, frame, valence... just to safeguard against malformed calls to this function? (Same comment to other symmetric "test_evaluate_blah" functions)

macedo22 replied:
How do you feel about me instead changing the test to only take one template parameter for the (shared) `TensorIndexType` (which is `SpatialIndex` or `SpacetimeIndex`) and one (shared) valence? Meaning, change
```
template <
    typename DataType, template <size_t, UpLo, typename> class TensorIndexTypeA,
    template <size_t, UpLo, typename> class TensorIndexTypeB, UpLo ValenceA,
    UpLo ValenceB, typename TensorIndexA, typename TensorIndexB>
void test_evaluate_rank_2_symmetric(
    const TensorIndexA& tensorindex_a,
    const TensorIndexB& tensorindex_b)
```
to
```
template <
    typename DataType, template <size_t, UpLo, typename> class TensorIndexType,
    UpLo Valence, typename TensorIndexA, typename TensorIndexB>
void test_evaluate_rank_2_symmetric(
    const TensorIndexA& tensorindex_a,
    const TensorIndexB& tensorindex_b)
```

As far as ensuring that `test_evaluate_rank_2_impl` is given the same (1) `TensorIndexType`s (`SpatialIndex` or `SpacetimeIndex` ) that also have the same (2) dimension, (3) valence, and (4) frame type, I  think (1) and (3) would be solved by the above, and (2) and (4) are already addressed since the `index_list` created uses the same dimension and frame type for both indices since they both use `DIM(data)` and `FRAME(data)` (see [here](https://github.com/sxs-collaboration/spectre/blob/b8a87e8ebcbd357ec61a553cd8defaed5cdebfac/tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank2TestHelpers.hpp#L141))

macedo22 replied:
I guess the only complication with this approach is that I wouldn't be able to use `\copydoc` to copy around the documentation because the `\tparam`s would be different (this has more of an impact on the rank 3 test functions since there are multiple). So I guess the decision is, which is better:
1) Reduce the template parameters as shown above, but now can only copy around the `\brief` and `\details` (instead of being able to wholesale copy documentation with `\copydoc` since the `tparam`s will be different)
2) Just put in static asserts, keeping redundant template parameters that are checked for equality

Let me know if you have a preference!

fmahebert replied:
- Ah, my comment was actually intended to question whether the two arguments to the function are of the same type, but it escaped me that -- as you've suggested for other template parameters -- you can ensure that `tensorindex_a` and `tensorindex_b` have the same type by replacing the two template parameters `TensorIndexA` and `TensorIndexB` with a single (shared) `TensorIndex`. This would definitely be cleaner than using a bunch of `static_asserts`!
- What you've suggested, of _also_ reducing the other template parameters to further enforce the symmetry, sounds like a good change as well! In terms of the choice you're suggesting, I'd personally opt for cleaner code at the expense of more documentation, i.e., option 1, but both would certainly be just fine. Do what you like best!

macedo22 replied:
Thanks! I attempted doing the first option, but ran into an issue that conflicts with a [previous change request](https://github.com/sxs-collaboration/spectre/pull/2475#discussion_r488906630) from @nilsdeppe . When I do the swap I mentioned in the [above comment](https://github.com/sxs-collaboration/spectre/pull/2475#discussion_r499100807), I updated the documentation of `test_evaluate_rank_2_symmetric` from simply:
```
/// \copydoc test_evaluate_rank_2_no_symmetry()
```

to:
```
/// \ingroup TestingFrameworkGroup
/// \copybrief test_evaluate_rank_2_no_symmetry()
///
/// \copydetails test_evaluate_rank_2_no_symmetry()
///
/// \tparam DataType the type of data being stored in the Tensors
/// \tparam TensorIndexType the shared \ref SpacetimeIndex "TensorIndexType" of
/// the indices
/// \tparam Valence the shared valence of the indices
/// \param tensorindex_a the first TensorIndex used on the RHS of the
/// TensorExpression, e.g. `ti_a`
/// \param tensorindex_b the second TensorIndex used on the RHS of the
/// TensorExpression, e.g. `ti_B`
```

This change was made because the template parameters of this symmetric function are now no longer the same as those for `test_evaluate_rank_2_no_symmetry`. However, the problem I'm seeing when viewing the documentation that this outputs is that `\copydetails` copies not only the `\details` section over, but also the `\tparam`s and `\param`s (the documentation shows both `\tparam`s and `\param`s from both functions). So then, the naive solution would be to manually copy and paste over the `\details` like I had before, but then at that point, if I'm writing the `\tparam`s, `param`s, and `\details` manually, that just leaves the `\brief` being copied. And since the `\brief`s originally were the only thing different between the functions that tested different symmetries, I don't know how much is gained from copying it with `\copybrief` vs just manually writing the briefs, as I originally did, because everything else in the functions' documentation will have to be manually written anyway. All of this is a little more imporrtant for the rank 3 functions, because there are 5 of those that call the `impl` function, but only 2 for the rank 2 case - meaning, more documentation to copy around vs manually copy + tweak.

Let me know if you two have thoughts on this. I'm happy to do whatever! I'm just not sure how to reconcile this because I think that removing the duplicate template parameters is a smart update, but I don't know how to update the documentation accordingly without conflicting with Nils' previous change request of not manually copying and pasting redundant documentation.

macedo22 replied:
(note: the above refers to the behavior I'm seeing locally - I haven't yet pushed a commit with these changes)

fmahebert replied:
I see the problem... I've run into a similar issue once and I ended up simply not documenting the template parameters. I'm not proud, but it got around the problem 😛.

One idea might be to continue using "copydoc" per Nils' suggestion, keep the change we've been discussing of simplifying the template parameters, but simply accept that the copied docs aren't completely accurate for the symmetric case. By adding to the documentation a short "Note: the function for symmetric tensors has fewer template parameters", it should hopefully be easy to see how to map from the documentation of the general case to the code with fewer template parameters?

I think I like this option better than the alternatives (writing the docs out twice, or keeping the duplicate template parameters). What do you and @nilsdeppe think?

nilsdeppe replied:
Sounds fine to me. I haven't gone through the full discussion in detail, so this might not be relevant, but keep in mind that you don't need to document deduced template parameters since the argument from which they are deduced is usually documented.

macedo22 replied:
Thanks so much for the input and feedback! I went ahead and made this change locally. However, the behavior I'm seeing now is that `\copydoc` isn't copying over the documentation when I do something like

```
/// \copydoc test_evaluate_rank_2_no_symmetry()
template <
    typename DataType, template <size_t, UpLo, typename> class TensorIndexType,
    UpLo Valence, typename TensorIndexA, typename TensorIndexB>
void test_evaluate_rank_2_symmetric(
```

There is no documentation for `test_evaluate_rank_2_symmetric` generated at all, as far as I can tell. I'm thinking that maybe this is due to the fact that `test_evaluate_rank_2_no_symmetry` and `test_evaluate_rank_2_symmetric` now have different template parameters, so `\copydoc` isn't able to correctly find the function (link object) from which to copy documentation. I tried searching doxygen's website and GitHub issues to see if there was a way to specialize the copying with template parameters, but I haven't found anything. Do either of you know if this is possible? And if there isn't a way to use `\copydoc` anymore, do you have a preferred alternative option?

fmahebert replied:
I'm not familiar enough of with Doxygen to know the answers here. One alternative might be (I think the two functions are back-to-back in the code?) to try using Doxygen's "@" syntax to assign the documentation to the two functions, like this:
```
//@{
/// doxygen comment
template<...> void func1(...) {...}
template<...> void func2(...) {...}
//@}
```
But I'm not sure if that will work for this case with the different template parameters, because it might just be different syntax for the same underlying copy action, and might fail in the same cases.

macedo22 replied:
Hmm I might actually be doing something wrong on my end. I reverted my changes back to when all of the template parameters were the same and used `\copydoc` and I found the same problem. Something else is preventing `\copydoc` from working correctly and I'm going to look into it. Just wanted to make sure I shared this here before you or Nils went out of your way to think of a work-around.

macedo22 replied:
FWIW, I was able to fix this issue by prepending with `/// \ingroup TestingFrameworkGroup`, like so:
```
/// \ingroup TestingFrameworkGroup
/// \copydoc test_evaluate_rank_2_no_symmetry()
```

fmahebert commented on tests/Unit/DataStructures/Tensor/Expressions/Test_TensorExpressions.cpp:
```diff
@@ -7,11 +7,239 @@
+      DataVector, SpatialIndex, SpacetimeIndex, SpatialIndex, UpLo::Lo,
+      UpLo::Lo, UpLo::Lo>(ti_i, ti_f, ti_j);
+
+  // Rank 3: DataVector;
```
Add "second and third indices symmetric" as elsewhere? :)

macedo22 replied:
Done!

fmahebert commented on tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank4TestHelpers.hpp:
```diff
@@ -0,0 +1,241 @@
+template <typename DataType, typename RhsSymmetry,
+          typename RhsTensorIndexTypeList, typename TensorIndexA,
+          typename TensorIndexB, typename TensorIndexC, typename TensorIndexD>
+void test_evaluate_rank_4(const TensorIndexA& tensorindex_a,
```
[not a change request, I just want to better understand] Why not use GENERATE_INSTANTIATIONS here to iterate over dims and frames as you've done in the lower-rank cases? Is it to avoid an explosion in the number of test cases and therefore test time?

macedo22 replied:
I think it was mostly just because when I was talking with @nilsdeppe weeks ago about test cases, we felt that if we exhaustively tested rank 3 tensors, there wasn't much of a need to exhaustively test rank 4 tensors since most of the edge cases you can think of with reordering symmetries and indices are already addressed by exhaustively testing rank 3. But, of course, exhaustively testing rank 4 is safer and more thorough.

Let me know if you would like me to implement the same exhaustive testing for rank 4 that I did with the others!

fmahebert replied:
No, I agree - I think you're in the diminishing returns territory here, where the test has to get a lot more complicated to test any potential remain edge cases. I think you can leave the code unchanged.

However, it might be useful to add a short note to the documentation to make the intent clear? Perhaps something like (I'm mostly reusing your words here) "The rank-3 case is exhaustively tested, and should test most edge cases with reordered indices and symmetries. So we test the rank-4 case less exhaustively to keep the test manageable"

macedo22 replied:
Diminishing returns is a good way to describe it - that's exactly it. We didn't think there was too much to gain. I'll add a note! I am a supporter of making things clear.

fmahebert commented on tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateTestHelpers.hpp:
```diff
@@ -0,0 +1,10 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
```
Is this file necessary? Why not include the 5 headers directly in the test?

macedo22 replied:
No, it doesn't serve much of a purpose. I only added it so that the number of `#include`s in `Test_TensorExpressions.cpp` is condensed. In my next PR, there will be a similar suite of tests for a new function, and that will be another 5 includes on top of these 5 for testing `evaluate`. So, that was really the only reason!

Let me know if you would like me to remove this file and just include the 5 helper files directly in `Test_TensorExpressions.cpp`. I don't feel strongly about it either way

fmahebert replied:
If this were 100 includes, I would definitely say to keep the extra file, but for 5 includes (and even with 5 more in the PR), personally I think I'd prefer just putting the include lines directly in the main test? It simplifies the process of looking up where functions are defined if there are fewer includes of includes...

macedo22 replied:
I'll remove it! Less work on the reader for tracing back through documentation

macedo22 replied:
Done!

fmahebert commented on src/DataStructures/Tensor/Expressions/Evaluate.hpp:
```diff
@@ -8,35 +8,115 @@
+          typename RhsSymmetry, typename RhsTensorIndexTypeList,
+          size_t NumIndices = tmpl::size<RhsSymmetry>::value,
+          typename IndexSequence = std::make_index_sequence<NumIndices>>
+struct LhsTensor;
```
I found it slightly confusing that this is called `LhsTensor` because if I correctly understand the documentation and code, this isn't a tensor so much as it is a mapping between tensor indices. If that's correct, I'd suggest a rename to make the functionality more evident, perhaps something like `LhsTensorIndexMap` or `TensorIndexMap` or ...? Feel free to throw out other suggestions too if you agree with the general idea but don't like these particular names.

macedo22 replied:
Thanks! I had a similar reservation when I was deciding on the name. I'll rename!

macedo22 replied:
@fmahebert One thing I will mention is that this `LhsTensor` struct computes the LHS symmetry and list of `TensorIndexType`s (based off of the map that it internally computes). So, it already has the reordered info for the LHS tensor using the mapping it internally computes, rather than it just containing the mapping to later be used externally to determine the LHS symmetry and list of `TensorIndexType`s (if that makes sense). Or in other words, it isn't *used* as a map in the context of `evaluate`, but rather is used to directly store index information about the LHS tensor (based off of the mapping it internally makes).

So, in a way, I think it is also slightly misleading to call it an "index map" since it represents the LHS tensor symmetry and `TensorIndexType`s, even though it does internally compute a map to accomplish this. I'm happy to still rename it as `TensorIndexMap` or `LhsTensorIndexMap` if you still think this is best, but I wanted to make sure I clarified the role of `LhsTensor` in case the purpose/function of the struct was unclear and if this changed your mind about how you think the struct should be named. `LhsTensorIndexInfo` would be a very poor name, but that almost best describes what it is, because it contains partial information about the LHS tensor to be constructed at the end of `evaluate`. I've been trying to think of a better name but haven't thought of one yet

fmahebert replied:
I think I see what you're saying, and I agree my first suggestions aren't ideal. Here's some other options I'll throw out: `LhsTensorIndexHelper` which is in the same spirit as your `LhsTensorIndexInfo` but, personally, I like it a little better; `LhsTensorIndices` (but of course there's also the symmetry information); `LhsTensorSymmAndIndices` (maybe my favorite so far?)... do any of these appeal to you?

I guess if nothing sticks at this stage, I'll suggest you keep the name as it currently is and we don't waste too much of your time on this 🙂

macedo22 replied:
I think `LhsTensorSymmAndIndices` is the most descriptive, so I went with that!

fmahebert commented on src/DataStructures/Tensor/Expressions/Evaluate.hpp:
```diff
@@ -8,35 +8,115 @@
  * \ingroup TensorExpressionsGroup
- * \brief Evaluate a Tensor Expression with LHS indices set in the template
- * parameters
+ * \brief Evaluate a right hand side tensor expression to a tensor with the LHS
```
[nitpick] "right hand side" -> "RHS", for consistency with other briefs

macedo22 replied:
Done!

fmahebert commented on src/DataStructures/Tensor/Expressions/Evaluate.hpp:
```diff
@@ -8,35 +8,115 @@
+ * with the same index ordering, such as \f$L_{ab} = R_{ab}\f$, or different
+ * ordering, such as \f$L_{ba} = R_{ab}\f$.
+ *
+ * ### Usage
```
I suggest "Usage" -> "Usage example" or "Example use" to make it more evident that other kinds of tensors and tensor expressions work too

macedo22 replied:
Done!

fmahebert reviewed (commented):
Good documentation and heroic testing effort! Here are some small suggestions for clarity and my own understanding, but all these are pretty minor so feel free to disagree with anything.

macedo22 commented:
@fmahebert I think everything thus far has been addressed! Let me know if you have any comments or if you would like me to squash! And if I do squash, b8a87e8 is a result of squashing a couple commits, so would you like all 6 commits squashed or should I leave b8a87e8 and squash the 5 after it into a second commit?

fmahebert commented:
@macedo22 The fixups LGTM, I think you can squash everything down to a single commit. I see I was rather naive in advertising my requests as being minor... sorry about the false advertising and thanks for working through it all! :D

macedo22 commented:
@fmahebert I think they were important improvements beyond just a face lift, so thank you!

nilsdeppe reviewed (approved):
We can ignore the clang-tidy error. That can get fixed later
